cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -61,217 +53,127 @@ type MachineTypeKey struct {
-			gc.migs[newMig.GceRef()].Config = newMig
+		if !reflect.DeepEqual(oldMig, newMig) {
+			gc.migs[newMig.GceRef()] = newMig
 			klog.V(4).Infof("Updated Mig %s", newMig.GceRef().String())
 			return true
 		}
 		return false
 	}
 
 	klog.V(1).Infof("Registering %s", newMig.GceRef().String())
-	// TODO(aleksandra-malinowska): fetch and set MIG basename here.
-	newMigInformation := &MigInformation{
-		Config: newMig,
-	}
-	gc.migs[newMig.GceRef()] = newMigInformation
+	gc.migs[newMig.GceRef()] = newMig
 	return true
 }
 
 // UnregisterMig returns true if the node group has been removed, and false if it was already missing from cache.
 func (gc *GceCache) UnregisterMig(toBeRemoved Mig) bool {
 	gc.cacheMutex.Lock()
 	defer gc.cacheMutex.Unlock()
-	gc.migsMutex.Lock()
-	defer gc.migsMutex.Unlock()
 
 	_, found := gc.migs[toBeRemoved.GceRef()]
 	if found {
 		klog.V(1).Infof("Unregistered Mig %s", toBeRemoved.GceRef().String())
 		delete(gc.migs, toBeRemoved.GceRef())
-		gc.removeInstancesForMig(toBeRemoved.GceRef())
+		gc.invalidateMigInstancesNoLock(toBeRemoved.GceRef())
 		return true
 	}
 	return false
 }
 
 // GetMigs returns a copy of migs list.
-func (gc *GceCache) GetMigs() []*MigInformation {
-	gc.migsMutex.Lock()
-	defer gc.migsMutex.Unlock()
+func (gc *GceCache) GetMigs() []Mig {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
 
-	migs := make([]*MigInformation, 0, len(gc.migs))
+	migs := make([]Mig, 0, len(gc.migs))
 	for _, mig := range gc.migs {
-		migs = append(migs, &MigInformation{
-			Basename: mig.Basename,
-			Config:   mig.Config,
-		})
+		migs = append(migs, mig)
 	}
 	return migs
 }
 
-// GetMigs returns a copy of migs list.
-func (gc *GceCache) getMigRefs() []GceRef {
-	gc.migsMutex.Lock()
-	defer gc.migsMutex.Unlock()
-
-	migRefs := make([]GceRef, 0, len(gc.migs))
-	for migRef := range gc.migs {
-		migRefs = append(migRefs, migRef)
-	}
-	return migRefs
-}
-
-func (gc *GceCache) updateMigBasename(migRef GceRef, basename string) {
-	gc.migsMutex.Lock()
-	defer gc.migsMutex.Unlock()
-
-	mig, found := gc.migs[migRef]
-	if found {
-		mig.Basename = basename
-	}
-	// TODO: is found == false a possiblity?
+// GetMig returns registered mig for given ref
+func (gc *GceCache) GetMig(migRef GceRef) (mig Mig, found bool) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+	mig, found = gc.migs[migRef]
+	return
 }
 
-// Methods locking on cacheMutex.
-
-// GetMigForInstance returns Mig to which the given instance belongs.
-// Attempts to regenerate cache if there is a Mig with matching prefix in migs list.
-// TODO(aleksandra-malinowska): reconsider failing when there's a Mig with
-// matching prefix, but instance doesn't belong to it.
-func (gc *GceCache) GetMigForInstance(instanceRef GceRef) (Mig, error) {
+// GetMigRefForInstance returns Mig to which the given instance belongs.
+func (gc *GceCache) GetMigRefForInstance(instanceRef GceRef) (migRef GceRef, found bool) {
```

Comment:
Can't we keep 'dumb' GetMigForInstance (which will only succeed if both MIG and instance are cached)? On cache miss, user can trigger regeneration, then try again.

Reply:
Not sure I'm following. Isn't the list of MIGs something we fetch from the API, repackage and cache on our side (in case of autodiscovery setup, and no-longer-here GKE cloud provider)?

Reply:
Mig. As per 'managed instance group` in GCE nomenclature yet. But not `Mig` object as defined in `gce` package. 
The former should be cached. And actually we do that by caching basename, targetsize, instance list - all of those are parts of managed instance group.
Yet `Mig` as a node-group which is subject to autoscaling by GCE cloud provider should IMO reside in gce_manager - and just use the cached bits from cache. Not be part of cache itself. 
Is my intention more clear?

Reply:
I mean caching list of MigRefs for which autoscaling is enabled, together with min/max/autoprovisioned.

Reply:
The `min/max/autoprovisioned` is not gce thing. Also there is no notion of "invalidating" of list of Migs. It still does not seem natural to hold `Migs` in GceCache.
(Though as I said I am not chaning it right now so we have some more time to think about it :) )